lib/amqp_client.js

```diff
@@ -157,20 +159,47 @@ AMQPClient.prototype.connect = function(url, policyOverrides) {
     self._connection.on(Connection.Connected, function(c) {
       debug('connected');
       self.emit(AMQPClient.ConnectionOpened);
+
+      var promises = [];
+      
+      // Only map the session if it has already been created
       if (self._session) {
         debug('session already exists, re-using');
         self._session.connection = self._connection;
-      } else {
-        self._session = self._newSession(c);
+
+        promises.push(new Promise(function(resolve, reject) {
+          self._session.once(Session.Mapped, function(s) {
+            debug('client session mapped');
+            resolve();
+          });
+
+          self._session.begin(self._session.policy || self.policy.session);
+        }));
       }
 
-      self._session.once(Session.Mapped, function(s) {
-        debug('mapped');
+      // Update any user-generated sessions
+      promises = promises.concat(self._userSessions.map(function(session) {
+        session.connection = self._connection;
+
+        return new Promise(function(resolve, reject) {
+          session.once(Session.Mapped, function(s) {
+            debug('user session mapped');
+            resolve();
+          });
+          
+          session.begin(session.policy || self.policy.session);
+        });
+      }));
+
+      if (promises.length > 0) {
+        Promise.all(promises).then(function() {
+          self.emit('connected');
```

Comment:
>self.emit('connected'); [](start = 10, length = 23)

unrelated but puzzling nonetheless to me: why would the "connected" state of the client be attached to sessions rather than connection or links?

Reply:
yeah this one is an artifact of how the Promise-based design works. it's trying to make sure things are in a usable state when the promise returns. since the session is baked into the connection by the library, that means both the connection and the session have to be ready. i considered making it lazy, but didn't know that i'd be able to get away with it without some subtle breaking changes

Reply:
actually i did make it lazy. the problem is reconnect because the session is already there